src/test/java/hudson/plugins/copyartifact/CopyArtifactWorkflowTest.java

```diff
@@ -146,4 +156,173 @@ private WorkflowJob createWorkflow(String name, String script) throws IOExceptio
+    // The latest pipeline supports calling `copyArtifacts` without `step`.
+
+    @Test
+    public void testDownstreamBuildSelector() throws Exception {
+        // upstream (freestyle) -> copiee (fresstyle)
+        // copier (pipeline) copies from copiee, which is downstream of `upstream`.
+        // DownstreamBuildSelector support detecting relations between only `AbstractProject`s.
+
+        FreeStyleProject upstream = jenkinsRule.createFreeStyleProject("upstream");
+        upstream.getBuildersList().add(new FileWriteBuilder("upstream_artifact.txt", "${BUILD_TAG}"));
+        ArtifactArchiver aa = new ArtifactArchiver("upstream_artifact.txt");
+        aa.setFingerprint(true);        // important to have Jenkins track builds
+        upstream.getPublishersList().add(aa);
+
+        FreeStyleProject copiee = jenkinsRule.createFreeStyleProject("copiee");
+        CopyArtifact ca = new CopyArtifact("upstream");
+        ca.setFingerprintArtifacts(true);       // important to have Jenkins track builds
+        copiee.getBuildersList().add(ca);
+        copiee.getBuildersList().add(new FileWriteBuilder("artifact.txt", "foobar"));
+        copiee.getPublishersList().add(new ArtifactArchiver("artifact.txt"));
+
+        jenkinsRule.assertBuildStatusSuccess(upstream.scheduleBuild2(0));
+        jenkinsRule.assertBuildStatusSuccess(copiee.scheduleBuild2(0));
+
+        WorkflowJob copier = createWorkflow(
+            "copier",
+            "step(copyArtifacts(projectName: 'copiee', selector: downstream(upstreamProjectName: 'upstream', upstreamBuildNumber: '1')));"
+            + "echo readFile('artifact.txt');"
+        );
+        jenkinsRule.assertLogContains("foobar", jenkinsRule.assertBuildStatusSuccess(copier.scheduleBuild2(0)));
+    }
+
+    @Test
+    public void testLastCompletedBuildSelector() throws Exception {
+        WorkflowJob copiee = createWorkflow(
+            "copiee",
+            "writeFile text: 'foobar', file: 'artifact.txt';"
+            + "archive includes: 'artifact.txt';"
+        );
+        jenkinsRule.assertBuildStatusSuccess(copiee.scheduleBuild2(0));
+
+        WorkflowJob copier = createWorkflow(
+            "copier",
+            "step(copyArtifacts(projectName: 'copiee', selector: lastCompleted()));"
+            + "echo readFile('artifact.txt');"
+        );
+        jenkinsRule.assertLogContains("foobar", jenkinsRule.assertBuildStatusSuccess(copier.scheduleBuild2(0)));
+    }
+
+    @Test
+    public void testParameterizedBuildSelector() throws Exception {
+        WorkflowJob copiee = createWorkflow(
+            "copiee",
+            "writeFile text: 'foobar', file: 'artifact.txt';"
+            + "archive includes: 'artifact.txt';"
+        );
+        jenkinsRule.assertBuildStatusSuccess(copiee.scheduleBuild2(0));
+
+        WorkflowJob copier = createWorkflow(
+            "copier",
+            "step(copyArtifacts(projectName: 'copiee', selector: buildParameter('<StatusBuildSelector/>')));"
```

Comment:
What?! Why is this taking an XML argument?

Reply:
`ParameterizedBuildSelector` takes a parameter created by `BuildSelectorParameter` which is a Xstream expression of a selector.
But this test is confusing and I'll update it to recieve a build parameter.